sdk/attestation/azure-security-attestation/inc/azure/attestation/attestation_client.hpp

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <azure/core/internal/json/json.hpp>
+#include <shared_mutex>
+#include <string>
+
+#include "attestation_client_options.hpp"
+#include <azure/attestation/attestation_client_models.hpp>
+#include <azure/core/internal/http/pipeline.hpp>
+
+namespace Azure { namespace Security { namespace Attestation {
+
+  using namespace Azure::Core;
+  class AttestationClient final {
+  private:
+    Azure::Core::Url m_endpoint;
+    std::string m_apiVersion;
+    std::shared_ptr<Azure::Core::Http::_internal::HttpPipeline> m_pipeline;
+
+    std::shared_timed_mutex m_sharedStateLock;
+    std::vector<AttestationSigner> m_attestationSigners;
+
+    void CacheAttestationSigners(Azure::Core::Context const& context);
+
+  public:
+    /**
+     * @brief Destructor.
+     *
+     */
+    virtual ~AttestationClient() = default;
+
+    /**
+     * @brief Construct a new Attestation Client object
+     *
+     * @param endpoint The URL address where the client will send the requests to.
+     * @param credential OPTIONAL The authentication method to use (required for TPM attestation).
+     * @param options The options to customize the client behavior.
+     */
+    explicit AttestationClient(
+        std::string const& endpoint,
+        std::shared_ptr<Core::Credentials::TokenCredential const> credential,
+        AttestationClientOptions options = AttestationClientOptions());
+
+    explicit AttestationClient(
+        std::string const& endpoint,
+        AttestationClientOptions options = AttestationClientOptions())
+        : AttestationClient(endpoint, nullptr, options)
+    {
+    }
+
+    std::string ClientVersion() const;
+    /**
+     * @brief Construct a new Key Client object from another key client.
+     *
+     * @param attestationClient An existing attestation client.
+     */
+    explicit AttestationClient(AttestationClient const& attestationClient)
+        : m_endpoint(attestationClient.m_endpoint), m_apiVersion(attestationClient.m_apiVersion),
+          m_pipeline(attestationClient.m_pipeline){};
+
+    /**
+     * @brief Retrieve the metadata signing certificates for this attestation instance.
+     *
+     * @returns Attestation Metadata.
+     */
+    Response<AttestationOpenIdMetadata> GetOpenIdMetadata(
+        Azure::Core::Context const& context = Azure::Core::Context::ApplicationContext);
```

Comment:
If the API won't change the client, it should be const as:
```suggestion
    Response<AttestationOpenIdMetadata> GetOpenIdMetadata(
        Azure::Core::Context const& context = Azure::Core::Context::ApplicationContext) const;
```

Add `const` to all API methods which won't mutate the client

Reply:
Most of my API methods will mutate the client,. But adding const anyway.